File: js/src/core/engine-adapter.js
/**
 * Engine Adapter - Abstract away Playwright/Puppeteer differences
 *
 * This module implements the Adapter pattern to encapsulate engine-specific
 * logic in a single place, following the "Protected Variations" principle.
 *
 * Benefits:
 * - Eliminates scattered `if (engine === 'playwright')` checks
 * - Easier to add new engines (e.g., Selenium)
 * - Easier to test with mock adapters
 * - Clearer separation of concerns
 */

import { TIMING } from './constants.js';

/**
 * Base class defining the engine adapter interface
 * All engine-specific operations should be defined here
 */
export class EngineAdapter {
  constructor(page) {
    this.page = page;
  }

  /**
   * Get engine name
   * @returns {string} - 'playwright' or 'puppeteer'
   */
  getEngineName() {
    throw new Error('getEngineName() must be implemented by subclass');
  }

  // ============================================================================
  // Element Selection and Locators
  // ============================================================================

  /**
   * Create a locator/element handle from a selector
   * @param {string} selector - CSS selector
   * @returns {Object} - Locator (Playwright) or ElementHandle (Puppeteer)
   */
  createLocator(selector) {
    throw new Error('createLocator() must be implemented by subclass');
  }

  /**
   * Query single element
   * @param {string} selector - CSS selector
   * @returns {Promise<Object|null>} - Locator/Element or null
   */
  async querySelector(selector) {
    throw new Error('querySelector() must be implemented by subclass');
  }

  /**
   * Query all elements
   * @param {string} selector - CSS selector
   * @returns {Promise<Array>} - Array of locators/elements
   */
  async querySelectorAll(selector) {
    throw new Error('querySelectorAll() must be implemented by subclass');
  }

  /**
   * Wait for selector to appear
   * @param {string} selector - CSS selector
   * @param {Object} options - Wait options {visible, timeout}
   * @returns {Promise<void>}
   */
  async waitForSelector(selector, options = {}) {
    throw new Error('waitForSelector() must be implemented by subclass');
  }

  /**
   * Wait for element to be visible
   * @param {Object} locatorOrElement - Locator or element
   * @param {number} timeout - Timeout in ms
   * @returns {Promise<Object>} - The locator/element
   */
  async waitForVisible(locatorOrElement, timeout = TIMING.DEFAULT_TIMEOUT) {
    throw new Error('waitForVisible() must be implemented by subclass');
  }

  /**
   * Count matching elements
   * @param {string} selector - CSS selector
   * @returns {Promise<number>} - Number of matching elements
   */
  async count(selector) {
    throw new Error('count() must be implemented by subclass');
  }

  // ============================================================================
  // Element Evaluation and Properties
  // ============================================================================

  /**
   * Evaluate function on element
   * @param {Object} locatorOrElement - Locator or element
   * @param {Function} fn - Function to evaluate
   * @param {any} args - Arguments to pass (optional)
   * @returns {Promise<any>} - Result of evaluation
   */
  async evaluateOnElement(locatorOrElement, fn, args) {
    throw new Error('evaluateOnElement() must be implemented by subclass');
  }

  /**
   * Get element text content
   * @param {Object} locatorOrElement - Locator or element
   * @returns {Promise<string|null>} - Text content
   */
  async getTextContent(locatorOrElement) {
    throw new Error('getTextContent() must be implemented by subclass');
  }

  /**
   * Get input value
   * @param {Object} locatorOrElement - Locator or element
   * @returns {Promise<string>} - Input value
   */
  async getInputValue(locatorOrElement) {
    throw new Error('getInputValue() must be implemented by subclass');
  }

  /**
   * Get element attribute
   * @param {Object} locatorOrElement - Locator or element
   * @param {string} attribute - Attribute name
   * @returns {Promise<string|null>} - Attribute value
   */
  async getAttribute(locatorOrElement, attribute) {
    throw new Error('getAttribute() must be implemented by subclass');
  }

  // ============================================================================
  // Element Interactions
  // ============================================================================

  /**
   * Click element
   * @param {Object} locatorOrElement - Locator or element
   * @param {Object} options - Click options {force, etc.}
   * @returns {Promise<void>}
   */
  async click(locatorOrElement, options = {}) {
    throw new Error('click() must be implemented by subclass');
  }

  /**
   * Type text into element (simulates typing)
   * @param {Object} locatorOrElement - Locator or element
   * @param {string} text - Text to type
   * @returns {Promise<void>}
   */
  async type(locatorOrElement, text) {
    throw new Error('type() must be implemented by subclass');
  }

  /**
   * Fill element with text (direct value assignment)
   * @param {Object} locatorOrElement - Locator or element
   * @param {string} text - Text to fill
   * @returns {Promise<void>}
   */
  async fill(locatorOrElement, text) {
    throw new Error('fill() must be implemented by subclass');
  }

  /**
   * Focus element
   * @param {Object} locatorOrElement - Locator or element
   * @returns {Promise<void>}
   */
  async focus(locatorOrElement) {
    throw new Error('focus() must be implemented by subclass');
  }

  // ============================================================================
  // Page-level Operations
  // ============================================================================

  /**
   * Evaluate JavaScript in page context
   * @param {Function} fn - Function to evaluate
   * @param {Array} args - Arguments to pass
   * @returns {Promise<any>} - Result of evaluation
   */
  async evaluateOnPage(fn, args = []) {
    throw new Error('evaluateOnPage() must be implemented by subclass');
  }

  /**
   * Get the main frame
   * @returns {Object} - Main frame
   */
  getMainFrame() {
    throw new Error('getMainFrame() must be implemented by subclass');
  }
}

/**
 * Playwright adapter implementation
 */
export class PlaywrightAdapter extends EngineAdapter {
  getEngineName() {
    return 'playwright';
  }

  // ============================================================================
  // Element Selection and Locators
  // ============================================================================

  createLocator(selector) {
    // Handle :nth-of-type() pseudo-selectors which don't work in Playwright locators
    const nthOfTypeMatch = selector.match(/^(.+):nth-of-type\((\d+)\)$/);
    if (nthOfTypeMatch) {
      const baseSelector = nthOfTypeMatch[1];
      const index = parseInt(nthOfTypeMatch[2], 10) - 1; // Convert to 0-based
      return this.page.locator(baseSelector).nth(index);
    }
    return this.page.locator(selector);
  }

  async querySelector(selector) {
    const locator = this.createLocator(selector).first();
    const count = await locator.count();
    return count > 0 ? locator : null;
  }

  async querySelectorAll(selector) {
    const locator = this.createLocator(selector);
    const count = await locator.count();
    const elements = [];
    for (let i = 0; i < count; i++) {
      elements.push(locator.nth(i));
    }
    return elements;
  }

  async waitForSelector(selector, options = {}) {
    const { visible = true, timeout = 5000 } = options;
    const locator = this.createLocator(selector);
    await locator.waitFor({ state: visible ? 'visible' : 'attached', timeout });
  }

  async waitForVisible(locatorOrElement, timeout = TIMING.DEFAULT_TIMEOUT) {
    const firstLocator = locatorOrElement.first();
    await firstLocator.waitFor({ state: 'visible', timeout });
    return firstLocator;
  }

  async count(selector) {
    return await this.page.locator(selector).count();
  }

  // ============================================================================
  // Element Evaluation and Properties
  // ============================================================================

  async evaluateOnElement(locatorOrElement, fn, args) {
    // Playwright only accepts a single argument
    if (args === undefined) {
      return await locatorOrElement.evaluate(fn);
    }
    return await locatorOrElement.evaluate(fn, args);
  }

  async getTextContent(locatorOrElement) {
    return await locatorOrElement.textContent();
  }

  async getInputValue(locatorOrElement) {
    return await locatorOrElement.inputValue();
  }

  async getAttribute(locatorOrElement, attribute) {
    return await locatorOrElement.getAttribute(attribute);
  }

  // ============================================================================
  // Element Interactions
  // ============================================================================

  async click(locatorOrElement, options = {}) {
    await locatorOrElement.click(options);
  }

  async type(locatorOrElement, text) {
    await locatorOrElement.type(text);
  }

  async fill(locatorOrElement, text) {
    await locatorOrElement.fill(text);
  }

  async focus(locatorOrElement) {
    await locatorOrElement.focus();
  }

  // ============================================================================
  // Page-level Operations
  // ============================================================================

  async evaluateOnPage(fn, args = []) {
    // Playwright only accepts a single argument (can be array/object)
    // To match Puppeteer's behavior where args are spread, we wrap the function
    // and pass all args as a single array, then apply them in the browser context
    if (args.length === 0) {
      return await this.page.evaluate(fn);
    } else if (args.length === 1) {
      return await this.page.evaluate(fn, args[0]);
    } else {
      // Multiple args - wrap function to accept array and spread them
      // This makes Playwright behave like Puppeteer's spread behavior
      // We pass the function string and args array, then reconstruct and call in browser
      const fnString = fn.toString();
      return await this.page.evaluate(
        ({ fnStr, argsArray }) => {
          // Reconstruct the function in browser context and call with spread args
          const reconstructedFn = new Function(`return (${fnStr})`)();
          return reconstructedFn(...argsArray);
        },
        { fnStr: fnString, argsArray: args }
      );
    }
  }

  getMainFrame() {
    return this.page.mainFrame();
  }
}

/**
 * Puppeteer adapter implementation
 */
export class PuppeteerAdapter extends EngineAdapter {
  getEngineName() {
    return 'puppeteer';
  }

  // ============================================================================
  // Element Selection and Locators
  // ============================================================================

  createLocator(selector) {
    // Puppeteer doesn't have locators - just returns selector
    // The actual element will be queried when needed
    return selector;
  }

  async querySelector(selector) {
    return await this.page.$(selector);
  }

  async querySelectorAll(selector) {
    return await this.page.$$(selector);
  }

  async waitForSelector(selector, options = {}) {
    const { visible = true, timeout = 5000 } = options;
    await this.page.waitForSelector(selector, { visible, timeout });
  }

  async waitForVisible(locatorOrElement, timeout = TIMING.DEFAULT_TIMEOUT) {
    // For Puppeteer, locatorOrElement is already an ElementHandle
    // We can't wait on it directly, so we just return it
    // The caller should have already used waitForSelector
    return locatorOrElement;
  }

  async count(selector) {
    const elements = await this.page.$$(selector);
    return elements.length;
  }

  // ============================================================================
  // Element Evaluation and Properties
  // ============================================================================

  async evaluateOnElement(locatorOrElement, fn, args) {
    // Puppeteer accepts the element as first arg, then spread args
    if (args === undefined) {
      return await this.page.evaluate(fn, locatorOrElement);
    }
    return await this.page.evaluate(fn, locatorOrElement, args);
  }

  async getTextContent(locatorOrElement) {
    return await this.page.evaluate((el) => el.textContent, locatorOrElement);
  }

  async getInputValue(locatorOrElement) {
    return await this.page.evaluate((el) => el.value, locatorOrElement);
  }

  async getAttribute(locatorOrElement, attribute) {
    return await this.page.evaluate(
      (el, attr) => el.getAttribute(attr),
      locatorOrElement,
      attribute
    );
  }

  // ============================================================================
  // Element Interactions
  // ============================================================================

  async click(locatorOrElement, options = {}) {
    await locatorOrElement.click(options);
  }

  async type(locatorOrElement, text) {
    // Puppeteer requires focus before typing
    await locatorOrElement.focus();
    await this.page.keyboard.type(text);
  }

  async fill(locatorOrElement, text) {
    // Puppeteer doesn't have fill() - use evaluate to set value
    await this.page.evaluate(
      (el, value) => {
        el.value = value;
        el.dispatchEvent(new Event('input', { bubbles: true }));
        el.dispatchEvent(new Event('change', { bubbles: true }));
      },
      locatorOrElement,
      text
    );
  }

  async focus(locatorOrElement) {
    await locatorOrElement.focus();
  }

  // ============================================================================
  // Page-level Operations
  // ============================================================================

  async evaluateOnPage(fn, args = []) {
    // Puppeteer accepts spread arguments
    return await this.page.evaluate(fn, ...args);
  }

  getMainFrame() {
    return this.page.mainFrame();
  }
}

/**
 * Factory function to create appropriate adapter
 * @param {Object} page - Playwright or Puppeteer page object
 * @param {string} engine - Engine type ('playwright' or 'puppeteer')
 * @returns {EngineAdapter} - Appropriate adapter instance
 */
export function createEngineAdapter(page, engine) {
  if (!page) {
    const errorDetails = {
      page,
      pageType: typeof page,
      engine,
      stackTrace: new Error().stack,
    };
    throw new Error(
      `page is required in createEngineAdapter. Received: page=${page} (type: ${typeof page}), engine=${engine}. This may indicate that the page object was not properly passed through the function call chain. Stack trace: ${errorDetails.stackTrace}`
    );
  }

  if (engine === 'playwright') {
    return new PlaywrightAdapter(page);
  } else if (engine === 'puppeteer') {
    return new PuppeteerAdapter(page);
  } else {
    throw new Error(
      `Unsupported engine: ${engine}. Expected 'playwright' or 'puppeteer'`
    );
  }
}

File: python/src/browser_commander/core/engine_adapter.py
"""Engine Adapter - Abstract away Playwright/Selenium differences.

This module implements the Adapter pattern to encapsulate engine-specific
logic in a single place, following the "Protected Variations" principle.

Benefits:
- Eliminates scattered `if engine == 'playwright'` checks
- Easier to add new engines
- Easier to test with mock adapters
- Clearer separation of concerns
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any

from browser_commander.core.constants import TIMING
from browser_commander.core.engine_detection import EngineType


class EngineAdapter(ABC):
    """Base class defining the engine adapter interface.

    All engine-specific operations should be defined here.
    """

    def __init__(self, page: Any) -> None:
        """Initialize the adapter with a page/driver object.

        Args:
            page: Playwright page or Selenium WebDriver
        """
        self.page = page

    @abstractmethod
    def get_engine_name(self) -> EngineType:
        """Get engine name.

        Returns:
            Engine type: 'playwright' or 'selenium'
        """
        ...

    # =========================================================================
    # Element Selection and Locators
    # =========================================================================

    @abstractmethod
    def create_locator(self, selector: str) -> Any:
        """Create a locator/element from a selector.

        Args:
            selector: CSS selector

        Returns:
            Locator (Playwright) or WebElement (Selenium)
        """
        ...

    @abstractmethod
    async def query_selector(self, selector: str) -> Any | None:
        """Query single element.

        Args:
            selector: CSS selector

        Returns:
            Locator/Element or None
        """
        ...

    @abstractmethod
    async def query_selector_all(self, selector: str) -> list[Any]:
        """Query all elements.

        Args:
            selector: CSS selector

        Returns:
            List of locators/elements
        """
        ...

    @abstractmethod
    async def wait_for_selector(
        self,
        selector: str,
        visible: bool = True,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> None:
        """Wait for selector to appear.

        Args:
            selector: CSS selector
            visible: Wait for visibility
            timeout: Timeout in milliseconds
        """
        ...

    @abstractmethod
    async def wait_for_visible(
        self,
        locator_or_element: Any,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> Any:
        """Wait for element to be visible.

        Args:
            locator_or_element: Locator or element
            timeout: Timeout in milliseconds

        Returns:
            The locator/element
        """
        ...

    @abstractmethod
    async def count(self, selector: str) -> int:
        """Count matching elements.

        Args:
            selector: CSS selector

        Returns:
            Number of matching elements
        """
        ...

    # =========================================================================
    # Element Evaluation and Properties
    # =========================================================================

    @abstractmethod
    async def evaluate_on_element(
        self,
        locator_or_element: Any,
        script: str,
        *args: Any,
    ) -> Any:
        """Evaluate JavaScript on element.

        Args:
            locator_or_element: Locator or element
            script: JavaScript code to evaluate
            *args: Arguments to pass to the script

        Returns:
            Result of evaluation
        """
        ...

    @abstractmethod
    async def get_text_content(self, locator_or_element: Any) -> str | None:
        """Get element text content.

        Args:
            locator_or_element: Locator or element

        Returns:
            Text content or None
        """
        ...

    @abstractmethod
    async def get_input_value(self, locator_or_element: Any) -> str:
        """Get input value.

        Args:
            locator_or_element: Locator or element

        Returns:
            Input value
        """
        ...

    @abstractmethod
    async def get_attribute(
        self,
        locator_or_element: Any,
        attribute: str,
    ) -> str | None:
        """Get element attribute.

        Args:
            locator_or_element: Locator or element
            attribute: Attribute name

        Returns:
            Attribute value or None
        """
        ...

    # =========================================================================
    # Element Interactions
    # =========================================================================

    @abstractmethod
    async def click(
        self,
        locator_or_element: Any,
        force: bool = False,
    ) -> None:
        """Click element.

        Args:
            locator_or_element: Locator or element
            force: Force click without checks
        """
        ...

    @abstractmethod
    async def type_text(self, locator_or_element: Any, text: str) -> None:
        """Type text into element (simulates typing).

        Args:
            locator_or_element: Locator or element
            text: Text to type
        """
        ...

    @abstractmethod
    async def fill(self, locator_or_element: Any, text: str) -> None:
        """Fill element with text (direct value assignment).

        Args:
            locator_or_element: Locator or element
            text: Text to fill
        """
        ...

    @abstractmethod
    async def focus(self, locator_or_element: Any) -> None:
        """Focus element.

        Args:
            locator_or_element: Locator or element
        """
        ...

    # =========================================================================
    # Page-level Operations
    # =========================================================================

    @abstractmethod
    async def evaluate_on_page(self, script: str, *args: Any) -> Any:
        """Evaluate JavaScript in page context.

        Args:
            script: JavaScript code to evaluate
            *args: Arguments to pass to the script

        Returns:
            Result of evaluation
        """
        ...

    @abstractmethod
    def get_url(self) -> str:
        """Get current page URL.

        Returns:
            Current URL
        """
        ...

    @abstractmethod
    async def goto(
        self,
        url: str,
        timeout: int = TIMING["NAVIGATION_TIMEOUT"],
    ) -> None:
        """Navigate to URL.

        Args:
            url: URL to navigate to
            timeout: Timeout in milliseconds
        """
        ...


class PlaywrightAdapter(EngineAdapter):
    """Playwright adapter implementation."""

    def get_engine_name(self) -> EngineType:
        """Get engine name."""
        return "playwright"

    def create_locator(self, selector: str) -> Any:
        """Create a Playwright locator."""
        # Handle :nth-of-type() pseudo-selectors
        import re

        match = re.match(r"^(.+):nth-of-type\((\d+)\)$", selector)
        if match:
            base_selector = match.group(1)
            index = int(match.group(2)) - 1  # Convert to 0-based
            return self.page.locator(base_selector).nth(index)
        return self.page.locator(selector)

    async def query_selector(self, selector: str) -> Any | None:
        """Query single element."""
        locator = self.create_locator(selector).first
        count = await locator.count()
        return locator if count > 0 else None

    async def query_selector_all(self, selector: str) -> list[Any]:
        """Query all elements."""
        locator = self.create_locator(selector)
        count = await locator.count()
        return [locator.nth(i) for i in range(count)]

    async def wait_for_selector(
        self,
        selector: str,
        visible: bool = True,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> None:
        """Wait for selector to appear."""
        locator = self.create_locator(selector)
        state = "visible" if visible else "attached"
        await locator.wait_for(state=state, timeout=timeout)

    async def wait_for_visible(
        self,
        locator_or_element: Any,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> Any:
        """Wait for element to be visible."""
        first_locator = locator_or_element.first
        await first_locator.wait_for(state="visible", timeout=timeout)
        return first_locator

    async def count(self, selector: str) -> int:
        """Count matching elements."""
        return await self.page.locator(selector).count()

    async def evaluate_on_element(
        self,
        locator_or_element: Any,
        script: str,
        *args: Any,
    ) -> Any:
        """Evaluate JavaScript on element."""
        if args:
            return await locator_or_element.evaluate(script, args[0])
        return await locator_or_element.evaluate(script)

    async def get_text_content(self, locator_or_element: Any) -> str | None:
        """Get element text content."""
        return await locator_or_element.text_content()

    async def get_input_value(self, locator_or_element: Any) -> str:
        """Get input value."""
        return await locator_or_element.input_value()

    async def get_attribute(
        self,
        locator_or_element: Any,
        attribute: str,
    ) -> str | None:
        """Get element attribute."""
        return await locator_or_element.get_attribute(attribute)

    async def click(
        self,
        locator_or_element: Any,
        force: bool = False,
    ) -> None:
        """Click element."""
        await locator_or_element.click(force=force)

    async def type_text(self, locator_or_element: Any, text: str) -> None:
        """Type text into element."""
        await locator_or_element.type(text)

    async def fill(self, locator_or_element: Any, text: str) -> None:
        """Fill element with text."""
        await locator_or_element.fill(text)

    async def focus(self, locator_or_element: Any) -> None:
        """Focus element."""
        await locator_or_element.focus()

    async def evaluate_on_page(self, script: str, *args: Any) -> Any:
        """Evaluate JavaScript in page context."""
        if not args:
            return await self.page.evaluate(script)
        if len(args) == 1:
            return await self.page.evaluate(script, args[0])
        # Multiple args - pass as array
        return await self.page.evaluate(script, list(args))

    def get_url(self) -> str:
        """Get current page URL."""
        return self.page.url

    async def goto(
        self,
        url: str,
        timeout: int = TIMING["NAVIGATION_TIMEOUT"],
    ) -> None:
        """Navigate to URL."""
        await self.page.goto(url, timeout=timeout)


class SeleniumAdapter(EngineAdapter):
    """Selenium adapter implementation."""

    def get_engine_name(self) -> EngineType:
        """Get engine name."""
        return "selenium"

    def create_locator(self, selector: str) -> Any:
        """Create a Selenium element locator (returns the selector for later use)."""
        # Selenium doesn't have locators - just return the selector
        return selector

    async def query_selector(self, selector: str) -> Any | None:
        """Query single element."""
        from selenium.common.exceptions import NoSuchElementException
        from selenium.webdriver.common.by import By

        try:
            return self.page.find_element(By.CSS_SELECTOR, selector)
        except NoSuchElementException:
            return None

    async def query_selector_all(self, selector: str) -> list[Any]:
        """Query all elements."""
        from selenium.webdriver.common.by import By

        return self.page.find_elements(By.CSS_SELECTOR, selector)

    async def wait_for_selector(
        self,
        selector: str,
        visible: bool = True,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> None:
        """Wait for selector to appear."""
        from selenium.webdriver.common.by import By
        from selenium.webdriver.support import expected_conditions as EC
        from selenium.webdriver.support.ui import WebDriverWait

        timeout_seconds = timeout / 1000
        wait = WebDriverWait(self.page, timeout_seconds)

        if visible:
            wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, selector)))
        else:
            wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, selector)))

    async def wait_for_visible(
        self,
        locator_or_element: Any,
        timeout: int = TIMING["DEFAULT_TIMEOUT"],
    ) -> Any:
        """Wait for element to be visible."""
        from selenium.webdriver.support import expected_conditions as EC
        from selenium.webdriver.support.ui import WebDriverWait

        timeout_seconds = timeout / 1000
        wait = WebDriverWait(self.page, timeout_seconds)
        return wait.until(EC.visibility_of(locator_or_element))

    async def count(self, selector: str) -> int:
        """Count matching elements."""
        elements = await self.query_selector_all(selector)
        return len(elements)

    async def evaluate_on_element(
        self,
        locator_or_element: Any,
        script: str,
        *args: Any,
    ) -> Any:
        """Evaluate JavaScript on element."""
        # Wrap the script to work with Selenium's execute_script
        wrapped_script = f"return (function(el, ...args) {{ {script} }})(arguments[0], ...Array.from(arguments).slice(1))"
        return self.page.execute_script(wrapped_script, locator_or_element, *args)

    async def get_text_content(self, locator_or_element: Any) -> str | None:
        """Get element text content."""
        return locator_or_element.text

    async def get_input_value(self, locator_or_element: Any) -> str:
        """Get input value."""
        return locator_or_element.get_attribute("value") or ""

    async def get_attribute(
        self,
        locator_or_element: Any,
        attribute: str,
    ) -> str | None:
        """Get element attribute."""
        return locator_or_element.get_attribute(attribute)

    async def click(
        self,
        locator_or_element: Any,
        force: bool = False,
    ) -> None:
        """Click element."""
        if force:
            # Use JavaScript click for force mode
            self.page.execute_script("arguments[0].click()", locator_or_element)
        else:
            locator_or_element.click()

    async def type_text(self, locator_or_element: Any, text: str) -> None:
        """Type text into element."""
        locator_or_element.send_keys(text)

    async def fill(self, locator_or_element: Any, text: str) -> None:
        """Fill element with text."""
        locator_or_element.clear()
        locator_or_element.send_keys(text)

    async def focus(self, locator_or_element: Any) -> None:
        """Focus element."""
        self.page.execute_script("arguments[0].focus()", locator_or_element)

    async def evaluate_on_page(self, script: str, *args: Any) -> Any:
        """Evaluate JavaScript in page context."""
        wrapped_script = f"return (function(...args) {{ {script} }})(...arguments)"
        return self.page.execute_script(wrapped_script, *args)

    def get_url(self) -> str:
        """Get current page URL."""
        return self.page.current_url

    async def goto(
        self,
        url: str,
        timeout: int = TIMING["NAVIGATION_TIMEOUT"],
    ) -> None:
        """Navigate to URL."""
        self.page.set_page_load_timeout(timeout / 1000)
        self.page.get(url)


def create_engine_adapter(page: Any, engine: EngineType) -> EngineAdapter:
    """Factory function to create appropriate adapter.

    Args:
        page: Playwright page or Selenium WebDriver object
        engine: Engine type ('playwright' or 'selenium')

    Returns:
        Appropriate adapter instance

    Raises:
        ValueError: If page is None or engine is unsupported
    """
    if page is None:
        msg = "page is required in create_engine_adapter"
        raise ValueError(msg)

    if engine == "playwright":
        return PlaywrightAdapter(page)
    if engine == "selenium":
        return SeleniumAdapter(page)

    msg = f"Unsupported engine: {engine}. Expected 'playwright' or 'selenium'"
    raise ValueError(msg)
